src/tactic/set.lean

```diff
@@ -0,0 +1,45 @@
+/-
+Copyright (c) 2021 Martin Zinkevich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Zinkevich
+-/
+import tactic.interactive
+import tactic.ext
+import tactic.tauto
+import data.set.default
+import data.finset.default
+
+
+/-! tactics for proving results about sets. Also work with finsets.
+
+set_taut proves set equality using intuitionistic logic.
+set_taut' proves set equality using classical logic.
+
+TODO: subset_taut: prove s ⊆ t.
+TODO: disjoint_taut: prove two sets are disjoint.
+TODO: unify all taut tactics into a single set tactic.
+TODO: set_simp: simplify sets. -/
+
+
+/-- tactic.tautology cannot deduce P b from (∀ b, P b ∧ Q b),
+  but it can from (∀ b, P b) ∧ (∀ b, Q b). -/
+lemma forall_and_iff_forall_and_forall {β} {P Q : β → Prop} :
+  (∀ b, P b ∧ Q b) ↔ (∀ b, P b) ∧ ∀ b, Q b :=
+iff.intro 
+  (λ h1 : (∀ b, P b ∧ Q b), and.intro (λ b, (h1 b).left) (λ b, (h1 b).right)) 
+  (λ h1 : (∀ b, P b) ∧ (∀ b, Q b), (λ b, and.intro (h1.left b) (h1.right b))) 
+
+/-- Helper for set_taut and set_taut'. -/
+meta def set_tauth (classical_op:bool) : tactic unit :=
+do tactic.ext1 [],
+do lemmas ← simp_lemmas.mk_default,
+do lemmas' ← simp_lemmas.add_simp lemmas `forall_and_iff_forall_and_forall,
+do tactic.simp_all lemmas' [],
+do tactic.tautology  {classical := classical_op}
```

Comment:
@mguaypaq is right that you don't need to repeat `do`.

However, without the classical switch (addressed in another comment), this is just a tactic macro. That is, it's shorthand for a sequence of lines you would write in begin-end blocks.

```suggestion
/-- Helper for set_taut and set_taut'. -/
meta def tactic.interactive.set_tauto : tactic unit :=
`[ext1, try {simp [forall_and_distrib]}; tauto!]
```

I added the `try` because `simp` may not rewrite anything, and the semicolon because `simp` may solve the goal on its own.

I think there's no need for the "helper" function. Interactive tactics (those that you use in begin-end blocks) should live in the `tactic.interactive` namespace.

Reply:
With this change, `set_tauto` will need a proper doc string. See the doc pages for examples -- it should explain what the tactic does and give a few examples of it in use. 

It will also need a tactic doc entry (https://leanprover-community.github.io/mathlib_docs/commands.html#add_tactic_doc).

Reply:
Alternatively, perhaps the tactic should take an optional list of additional simp lemmas. I'm not sure if/when this helps.

Reply:
I think an additional list of simp lemmas would help, I just don't know how to do that. :-)

Reply:
I simplified the example, added some trys. There is something happening in six examples that I don't understand.